### PR TITLE
[enhancement] add ```__dlpack__``` and ```__dlpack_device__``` attributes to python oneDAL table class

### PR DESCRIPTION
**Pull request by icfaust**
## Description

I missed something with respect to array_api support in scikit-learn, namely that the return types should match for stored attributed when ```array_api_dispatch``` is enabled.  This requires that the returned tables be convertible to array_api arrays rather than numpy arrays, which would be impossible for gpu inputs without dpctl at the current moment (requiring some significant complexity).  This enables the ```__dlpack__``` and ```__dlpack_device__``` attributes which allow for python oneDAL tables to be converted into array_api inputs via the ```from_dlpack``` method (https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.from_dlpack.html). A follow-up PR will use this new C++ infrastructure to change the backend function ```from_table``` to directly return the table into the proper array type (and will be the next-step PR).  This is necessary for true array_api support in matching sklearn.

Additional tests must be written for enabled algorithms to verify matching return types and stored attributes to the submitted data.

These attributes aren't used in any workflows (yet), no performance benchmarks necessary

The following changes occur in this PR:
 - Add ```__dlpack__``` attribute to pybind11 oneDAL table interface for generating a capsule of a DLManagedTensor (Not versioned because of low support in other frameworks (see pytorch source code))
 - Add ```__dlpack_device__``` attribute to pybind11 oneDAL table interface (requirement for dpctl's ```from_dlpack```)
 - Add DLPack enums (cpu, oneAPI) to pybind11 backends to interface with other frameworks

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/datatypes/tests/test_data.py [hunk 1]
```diff
@@ -574,3 +574,40 @@ def test_table_conversions_dlpack(dataframe, queue, order, data_shape, dtype):
...
+    X_df = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+
+    X_table = to_table(X_df)
+    if hasattr(X_df, "__dlpack_device__"):
+        assert X_df.__dlpack_device__() == X_table.__dlpack_device__()
+
+    if xp := getattr(X_df, "__array_namespace__", lambda: None)():
+        X_out = xp.from_dlpack(X_table)
+        X_temp = xp.asnumpy(X_out) if hasattr(xp, "asnumpy") else np.asarray(X)
+        assert_allclose(np.squeeze(X_temp), np.squeeze(X))
```
Comment (david-cortes-intel) [inline onedal/datatypes/tests/test_data.py:603]: Should't this test for exact equality if there's no dtype conversion?

File: onedal/datatypes/tests/test_data.py [hunk 2]
```diff
@@ -574,3 +574,40 @@ def test_table_conversions_dlpack(dataframe, queue, order, data_shape, dtype):
...
+
+    X_df = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+
+    X_table = to_table(X_df)
+    if hasattr(X_df, "__dlpack_device__"):
+        assert X_df.__dlpack_device__() == X_table.__dlpack_device__()
+
+    if xp := getattr(X_df, "__array_namespace__", lambda: None)():
+        X_out = xp.from_dlpack(X_table)
+        X_temp = xp.asnumpy(X_out) if hasattr(xp, "asnumpy") else np.asarray(X)
```
Comment (david-cortes-intel) [inline onedal/datatypes/tests/test_data.py:602]: Per the docstring at the top of this test, shouldn't it always check that `np.asarray` works?

File: onedal/datatypes/tests/test_data.py [hunk 3]
```diff
@@ -574,3 +574,40 @@ def test_table_conversions_dlpack(dataframe, queue, order, data_shape, dtype):
...
+        assert_allclose(np.squeeze(X_temp), np.squeeze(X))
+    else:
+        # only some numpy versions support array_api and from_dlpack
+        pytest.skip(f"{dataframe} does not have an __array_namespace__ attribute")
+
+    # test capsule deletion, should have no impact on underlying memory
+    # important for testing ``dlpack::free_capsule``
+    capsule = X_table.__dlpack__()
+    assert_allclose(np.squeeze(from_table(X_table)), np.squeeze(X))
+    del capsule
```
Comment (david-cortes-intel) [inline onedal/datatypes/tests/test_data.py:612]: I think there's no guarantee that the deletor will be called right at this point. Here you could try to forcibly trigger garbage collection:
```python
import gc
gc.collect()
```

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 4]
```diff
@@ -154,6 +154,105 @@ dal::table convert_to_table(py::object obj, py::object q_obj, bool recursed) {
     return res;
 }
 
+DLDevice get_dlpack_device(const dal::array<byte_t>& array) {
+    DLDevice device;
+#ifdef ONEDAL_DATA_PARALLEL
+    // std::optional<sycl::queue>
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp:160]: Forgot to remove this?

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 5]
```diff
@@ -154,6 +154,105 @@ dal::table convert_to_table(py::object obj, py::object q_obj, bool recursed) {
...
+    tensor.strides = &tensor.shape[2];
+    tensor.byte_offset = std::uint64_t(0);
+
+    return tensor;
+}
+
+static void free_capsule(PyObject* cap) {
+    DLManagedTensor* dlm = nullptr;
+    if (PyCapsule_IsValid(cap, "dltensor")) {
+        dlm = reinterpret_cast<DLManagedTensor*>(PyCapsule_GetPointer(cap, "dltensor"));
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp:214]: I think this should be a static cast.

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 6]
```diff
@@ -154,6 +154,105 @@ dal::table convert_to_table(py::object obj, py::object q_obj, bool recursed) {
...
+        auto stored_array = static_cast<dal::array<byte_t>*>(self->manager_ctx);
+        if (stored_array) {
+            delete stored_array;
+        }
+        delete[] self->dl_tensor.shape;
+        delete self;
+    };
+
+    // create capsule
+    py::capsule capsule(reinterpret_cast<void*>(dlm), "dltensor", free_capsule);
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp:252]: Also static cast.

File: onedal/datatypes/tests/test_data.py (hunk 2 above)
Comment (icfaust) [reply onedal/datatypes/tests/test_data.py:602]: Sorry, don't understand fully.  The ```__dlpack__``` table attribute is used on line 601, but it needs to be converted to numpy from these other frameworks for value testing with ```assert_allclose```. Line 602 is necessary for conversion from dpctl/dpnp to numpy, as np.asarray on dpnp/dpctl inputs will make very weird numpy object arrays rather than numeric ones.

File: onedal/datatypes/tests/test_data.py (hunk 1 above)
Comment (icfaust) [reply onedal/datatypes/tests/test_data.py:603]: At first glance I agree with your assessment, the only problem is this test on reduced precision hardware (i.e. fp32-only support), where a downconversion may occur from fp64 and then it wouldn't be an exact comparison.  If you think it should if statement for that case let me know, and I'll do it.

File: onedal/datatypes/tests/test_data.py (hunk 3 above)
Comment (icfaust) [reply onedal/datatypes/tests/test_data.py:612]: You are right, the only thing that let me know it was working was changes in the coverage afterward.  I'll do your suggestion as it is more guaranteed.

File: onedal/datatypes/tests/test_data.py (hunk 1 above)
Comment (david-cortes-intel) [reply onedal/datatypes/tests/test_data.py:605]: Then how about checking for exact equality when dtypes are the same, and approximate when not? Or how about in the latter case downcasting the numpy array and checking for exact equality on that?

File: onedal/datatypes/tests/test_data.py (hunk 2 above)
Comment (david-cortes-intel) [reply onedal/datatypes/tests/test_data.py:604]: Wouldn't that kind of behavior on their part (producing non-numeric numpy arrays) be non-compliant with the array API specification?

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 4 above)
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp]: removed

**Comment by icfaust** [comment]
@david-cortes-intel thanks for helping on this PR. let me know if its good to go, or if you'd like another reviewer to double check things.

**Review by david-cortes-intel — APPROVED** [review]
